### DISCUSSION TITLE
I'm trying to split the logic into different files, but it only works in main

### DISCUSSION THREAD

**Discussion by klaydev** [miguelgrinberg/python-socketio#1198] [General]
Hi! I am trying to split the logic from one file into several, here is an example of my code:
```
#app.py
from fastapi import FastAPI
from socketio import ASGIApp, AsyncServer

app = FastAPI()
sio = AsyncServer(async_mode="asgi")
sio_app = ASGIApp(sio, app)

app.mount('/', sio_app)

@sio.on("connect")
async def connect(sid, environ):
    print("connect", sid)

```
```
#ws.py
from app import sio


@sio.on("test")
async def test(sid): <-- ITS NOT WORKING
    print("test!")
    await sio.emit("test", "test!")
```
```
#client.py
from socketio import AsyncClient
import asyncio

client = AsyncClient()

async def main():
    await client.connect("http://127.0.0.1:8000")
    await asyncio.sleep(1)
    await client.emit("test", "1234")

@client.on("test")
async def test(t):
    print(t)


@client.on("connect")
async def connect():
    print("connected!")

if __name__ == "__main__":
    asyncio.run(main())
```

**Comment by miguelgrinberg** [comment]
What do you mean by "it's not working"? The ws.py file is not imported anywhere, maybe that is the problem? Also the way you are mounting Socket.IO into app, and app into Socket.IO is likely not what you want. You have to decide which of the two ASGI apps are going to be in the front, and only mount the other one under it.

**Reply by klaydev** [reply]
> What do you mean by "it's not working"? The ws.py file is not imported anywhere, maybe that is the problem? Also the way you are mounting Socket.IO into app, and app into Socket.IO is likely not what you want. You have to decide which of the two ASGI apps are going to be in the front, and only mount the other one under it.

By "it doesn't work" I mean that it really doesn't want to work. I've tried all the methods. Can you please tell me how I can fix this?

**Reply by miguelgrinberg** [reply]
I'm sorry, I don't think I'm making myself clear. You cannot just tell me "it doesn't work" and leave me with the task to figure out what in all the code that you provided is failing, and with what kind of error. If you provide all the details I can maybe find a pattern and suggest a solution. Without details it would require me too much time to set up an example based on the code you provided that I can debug. Feels unfair that you expect me to do all that work, when you can just describe the problem in full details.

**Reply by klaydev** [reply]
Sorry, I didn't put it that way. I'm trying to break the logic of working from one file into several, please tell me how I can do this? In the very first message, I gave an example of the code. Maybe I don't quite understand how it should work, but in my mind, if I import `sio` from `app.py `, then it should work.

**Reply by miguelgrinberg** [reply]
You are still not telling what is the problem with the code above. I understand it doesn't work. You need to provide details on the problem.

**Reply by klaydev** [reply]
So, I understand what you're talking about. The problem is that when I import `sio` from `app.py ` in `ws.py ` then the handler in `ws.py ` it doesn't work

**Reply by miguelgrinberg** [reply]
Where is ws.py imported? That's not in the code you provided.

**Reply by klaydev** [reply]
The code that I threw off is the weight that is present. It is correctly noted that `ws.py ` not imported anywhere. In order for this to work, do I need to import it somewhere? And if, let's say, I have 10 or 20 `ws.py ` with similar functionality? Is there any solution, if I understand correctly, to somehow force the program to automatically import these modules?

**Reply by miguelgrinberg** [reply]
There are lots of ways to import code in Python. But that is something that is outside the scope of the support I provide here and is more appropriate to ask in a Python support board.

**Reply by klaydev** [reply]
OK, thanks for the guidance!

**Comment by klaydev** [comment]
I solved the problem, but now there is another problem. Here is the code:
```py
#main.py
import importlib, os, socketio
from fastapi import FastAPI


app = FastAPI()

sio = socketio.AsyncServer(async_mode="asgi", cors_allowed_origins="*")

for root, dirs, files in os.walk("ws"):
    for file in files:
        if file.split('_')[-1][:-3] == "handler":
            module = importlib.import_module(root.replace("\\", ".") + '.' + file[:-3])
            module.setup(sio)

print(sio.namespaces)

sio_app = socketio.ASGIApp(sio, app)



app.add_route("/socket.io/", route=sio_app, methods=["GET", "POST"])
app.add_websocket_route("/socket.io/", sio_app)
```
```py
#ws/test_handler.py
import socketio


class TestHandler(socketio.AsyncNamespace):
    def on_connect(self, sid, environ):
        pass

    def on_disconnect(self, sid):
        pass

    async def on_test(self, sid):
        print("test", sid)


def setup(sio):
    sio.register_namespace(TestHandler("/test"))
```
```py
#ws/2/12_handler.py
import socketio


class TestNamespace(socketio.AsyncNamespace):
    def on_connect(self, sid, environ):
        print("Connect", sid)

    def on_disconnect(self, sid):
        print("disconnect", sid)


def setup(sio):
    sio.register_namespace(TestNamespace("/"))
```
As a result, print(sio.namespaces) outputs ['/'], but it should be ['/', '/test']

**Reply by klaydev** [reply]
@miguelgrinberg

**Reply by miguelgrinberg** [reply]
First of all, the `sio.namespaces` attribute is not a public attribute that is documented. Second, I'm wondering if you are 100% positive that your files are being imported. My guess is that you are still having Python issues and your code needs some debugging.

**Reply by klaydev** [reply]
Uh-huh, please tell me, can I make two namespaces with the same path? If not, then please hint that you can do something else, similar in functionality.

**Reply by miguelgrinberg** [reply]
I don't know what "same path" means in the context of a namespace. You have `/` and `/test`, so it's not the same. If you are asking if a single server instance can have more than one namespace, sure, there is no limit.
